Make `ColumnTransfomer` automatically convert sparse inputs to CSR format

AYUSH27112021 opened the pull request:
Fixes #30275

#### What does this implement/fix

Enhanced ` _check_X` to verify for sparse matrix subscriptability via `__getitem__`. Non-subscriptable sparse formats now raise a warning and are converted to CSR format for compatibility.

#### Any other comments?
The `_check_X ` function in `sklearn/compose/_column_transformer.py`, line 1324:

##### Current behavior:
Only checks for `sparse.issparse(X)` which returns True for dia_array(or any other sparse matrix type from scipy.sparse).

This causes a `TypeError: 'dia_array' object is not subscriptable` when non-subscriptable sparse formats are used and `fit_transform` is called.

##### Fix:
This update ensures that non-subscriptable sparse formats are handled by raising a warning and converting them to CSR format, preventing errors during transformation.

AYUSH27112021 committed:
Update _column_transformer.py

Enhanced _check_X to check for sparse matrix subscriptability via __getitem__. Non-subscriptable sparse formats (e.g., COO, DIA) now raise a warning and are converted to CSR format for compatibility.

AYUSH27112021 committed:
Update _column_transformer.py

Enhanced _check_X to check for sparse matrix subscriptability via __getitem__. Non-subscriptable sparse formats (e.g., COO, DIA) now raise a warning and are converted to CSR format for compatibility.

AYUSH27112021 committed:
Update _column_transformer.py

Fixed linting error

AYUSH27112021 committed:
Merge branch 'main' into main

AYUSH27112021 committed:
Fixed Linting issues

AYUSH27112021 committed:
Merge branch 'main' of https://github.com/AYUSH27112021/scikit-learn-Fixes-30275

ogrisel commented on sklearn/compose/_column_transformer.py:
```diff
@@ -1317,12 +1317,39 @@ def get_metadata_routing(self):
+                    "is not subscriptable."
+                    "It will be converted to CSR format for compatibility."
+                ),
+                category=UserWarning,
```
If we want a warning, we should rather use `sklearn.exceptions.DataConversionWarning`. But I am not sure if we should warn at all, as there is little the user can do except, manually convert to CSR themselves, which wouldn't be more efficient than what we do.

AYUSH27112021 replied:
I agree , we can remove the warning. However, we should include a message in the documentation or a note within the function itself. I am unsure which specific documentation needs to be updated. Could you let me know?

AYUSH27112021 committed:
Merge branch 'scikit-learn:main' into main

AYUSH27112021 committed:
Removed warning on CSR format

Removed the User warning added in this PR. Sparse inputs are automatically converted to the CSR format as this is currently the most efficient approach.

ogrisel commented on sklearn/compose/_column_transformer.py:
```diff
@@ -1317,12 +1317,31 @@ def get_metadata_routing(self):
+
+    Parameters
+    ----------
+    Xs : list of {array-like, sparse matrix, dataframe}
```
The input is a single data structure `X`, not a list of such datastructures named `Xs`.

ogrisel commented on sklearn/compose/_column_transformer.py:
```diff
@@ -1317,12 +1317,31 @@ def get_metadata_routing(self):
+
+    For other input types, such as DataFrame-like objects or array-like structures
+    with `__array__` and `shape` attributes, return them as-is without additional
+    validation to optimize performance.
```
I think the two above comments paragraph could be moved to inline comments where appropriate in the body of the function.

`_check_X` is private, so this docstring will not be rendered as part of the documentation, hence no need for a structured docstring.

On the other hand, we could add a note to the `X` parameter entry of the `fit` / `fit_transform` and/or `transform` docstring to let the user know that sparse inputs that are not row-wise indexable by default will be internally converted to the CSR format.

AYUSH27112021 replied:
@ogrisel I found that `check_array` already implements this functionality using the `accept_sparse=("csr", "csc", "lil", "dok")` So, I'll use that function instead for non-subscriptable sparse formats. 

While reading through the implementation, I noticed a TODO in the `_ensure_sparse_format` function (https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/utils/validation.py#L654-L660). Since SciPy's minimum requirement is 1.6, we can remove this as well.

AYUSH27112021 replied:
> On the other hand, we could add a note to the `X` parameter entry of the `fit` / `fit_transform` and/or `transform` docstring to let the user know that sparse inputs that are not row-wise indexable by default will be internally converted to the CSR format.

@ogrisel once you review these changes ill add a note in the docs

AYUSH27112021 replied:
Done changes to Docstring of `_fit`, `_fit_transform`

ogrisel reviewed (commented):
Please add a non-regression test based on the code snippet of the original issue under `sklearn/compose/tests/test_column_transformer.py` to check that it's now possible to use the column transformer on dia or COO formatted sparse matrix inputs without getting an error.

Please also add a changelog entry to document this fix under `doc/whats_new/upcoming_changes/sklearn.compose`.

AYUSH27112021 committed:
non-regression test added

1-  removed Docstring from `_check_X`.
2- added a non-regression unit test for verification.
3- added the changelog entry.

AYUSH27112021 committed:
Merge branch 'main' into main

AYUSH27112021 commented:
any updates on this PR

AYUSH27112021 committed:
Merge branch 'main' into main

AYUSH27112021 committed:
Merge branch 'scikit-learn:main' into main

AYUSH27112021 committed:
Added note about X in _fit and it_transform Docstring

AYUSH27112021 committed:
Merge branch 'main' into main

AYUSH27112021 committed:
Merge branch 'scikit-learn:main' into main

AYUSH27112021 commented:
Hi @ogrisel,
If this issue has already been resolved or is no longer needed, I’d like to go ahead and close this PR. If any changes are still required, please let me know. Otherwise, I’ll proceed with closing it.

Thanks!

thomasjpfan commented on sklearn/compose/_column_transformer.py:
```diff
@@ -980,6 +985,11 @@ def fit_transform(self, X, y=None, **params):
+        Notes
+        -----
+        If `X` is a sparse matrix and not row-wise indexable by default, it will
+        be internally converted to the CSR format for efficient processing.
```
Efficiency depends on different use cases:

 - If the inner transformers works on the CSC directly, then it's most efficient for the column transformer to use CSC.
 - If the inner transformers works on CSR, then this PR is the most efficient.

I'm okay with this PR using CSR, but I would remove the "efficient processing" in the docstring:

```suggestion
        be internally converted to the CSR format.
```

thomasjpfan commented on sklearn/compose/_column_transformer.py:
```diff
@@ -1333,12 +1343,16 @@ def __sklearn_tags__(self):
+
+    """
+    if sparse.issparse(X):
+        if not hasattr(X, "__getitem__"):
+            X = sparse.csr_matrix(X)
+        return X
```
Can we reuse `check_array` here?

```suggestion
        return check_array(X, accept_sparse=("csc", "csr"), ensure_all_finite=False)
```

AYUSH27112021 replied:
We could use `check_array`, which raises an error for non-indexable sparse formats. However, for performance reasons, I used `hasattr(X, "__getitem__")` to quickly check if the input is indexable.

My initial idea was to catch non-indexable inputs early, convert them to CSR format, and issue a warning to the user. However, I initially overlooked the differences between CSR and CSC formats.

Given that, we have the following possible paths:

Use `check_array` as you suggested.

Use` hasattr(X, "__getitem__")` and raise a error if the input is not indexable, let the user to correct it.

Automatically convert non-indexable sparse inputs to CSR (as done in a previous commit) and raise a warning to inform the user of the conversion.

Let me know which approach you prefer, and I'll format the changes accordingly.

thomasjpfan replied:
Performance wise, `check_array` is basically the same as `csr_matrix` if `ensure_all_finite=False`. Looking at `check_array`, we can not actually use it because of the complex numbers check: https://github.com/scikit-learn/scikit-learn/blob/00d3ef9f4d7e224e59f9e01f678abb918231858f/sklearn/utils/validation.py#L1011

---

The `hasattr(X, "__getitem__")` feels too implicit. (It'll be harder for future maintainers/contributors to understand "why") If we go for auto-conversion, I prefer something like this:

```python
if issparse(X):
    if X.format not in ("csr", "csc", "lil", "dok"): 
        return sparse.csr_matrix(X)
```

> Use hasattr(X, "__getitem__") and raise a error if the input is not indexable, let the user to correct it.

I do like throwing a better error message. If we do the auto `csr_matrix` conversion, I suspect someone would also want that option to convert it to `csc_matrix`, which complicates `ColumnTransformer` more.

AYUSH27112021 committed:
Update _column_transformer.py

AYUSH27112021 committed:
Merge branch 'main' into main

AYUSH27112021 committed:
Update _column_transformer.py

AYUSH27112021 committed:
Update _column_transformer.py

AYUSH27112021 commented:
Hi @thomasjpfan ,
I've followed your suggestions and made the necessary changes.
Could you please review them and let me know if any further adjustments are needed?

AYUSH27112021 committed:
Update _column_transformer.py

AYUSH27112021 committed:
Update _column_transformer.py

AYUSH27112021 committed:
Update _column_transformer.py

thomasjpfan reviewed (approved):
Thank you for the update! My preferences:

- I am +0.6 on doing the automatic conversion.
- I am +1 on raising a better error, telling the user that we only support a certain subset of sparse data.

I'll approve, and see what other think.

AYUSH27112021 committed:
Merge branch 'main' into main

AYUSH27112021 commented:
hi @thomasjpfan i guess @ogrisel is not active can you assign someone else

AYUSH27112021 commented:
Hi @thomasjpfan, I’m going to close this branch due to inactivity and lack of updates. Please feel free to reopen it or create a new Pull request.

Thanks!
